executors.md

```diff
@@ -0,0 +1,179 @@
+# Summary
+[summary]: #summary
+
+This RFC proposes a design for `futures-executors`, including both executor
+traits and built-in executors. In addition, it sets up a core expectation for all
+tasks that they are able to spawn additional tasks, while giving fine-grained
+control over what executor that spawning is routed to.
+
+NOTE: this RFC assumes that [RFC #2] is accepted.
+
+[RFC #2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+# Motivation
+[motivation]: #motivation
+
+This is a follow-up to the [Tokio Reform], [Futures 0.2], and [Task Context]
+RFCs, harmonizing their designs.
+
+[Tokio Reform]: https://github.com/tokio-rs/tokio-rfcs/pull/3
+[Futures 0.2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/1
+[Task Context]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+The design in this RFC has the following goals:
+
+- Add a core assumption that tasks are *always* able to spawn additional tasks,
+  avoiding the need to reflect this at the API level. (Note: this assumption is
+  only made in contexts that also assume an allocator.)
+
+- Provide fine-grained control over which executor is used to fulfill that assumption.
+
+- Provide reasonable built-in executors for both single-threaded and
+  multithreaded execution.
+
+It brings the futures library into line with the design principles of the [Tokio
+Reform], but also makes some improvements to the `current_thread` design.
+
+# Proposed design
+
+## The core executor abstraction: `Spawn`
+
+First, in `futures-core` we provide a basic definition of a general-purpose
+executor:
+
+```rust
+trait Spawn {
+    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>);
+}
+```
+
+The `Spawn` trait is quite straightforward (some alternatives and tradeoffs are
+discussed in the next section), and crucially is object-safe.
+
+We then build in an executor to the task context, stored internally as a trait
+object, which allows us to provide the following methods:
+
+```rust
+impl task::Context {
+    // Spawn onto the current default executor
+    fn spawn<F>(&self, F)
```

Comment:
Is there a reason this isn't `&mut self`? I've been finding that using `&self` ended up hitting problems in a number of cases. Also, since this is stored on `Context`, you have mutable access anyway.

`Spawn` instances that do require `&self` can always impl `Spawn for &Self`, like `TcpStream` and co. w/ `Read`.

Reply:
I'd prefer if this function were not part of the RFC (it's functionality should be possible with the `with_spawn` function, right?). I think it would be a good idea to get experience with people having to pick a default before specifying one, especially since specifying one creates friction when it needs to be changed.

Reply:
@ahmedcharles A major benefit of this RFC is to provide the guarantee that all `std`-only `Future` implementations have the ability to `spawn` subtasks. This removes the current separation that exists between functions that take an `Executor` handle and those that do not. When a deeply nested future wants to spawn tasks as part of its internal implementation, it currently needs to have an `Executor` handle threaded down through arguments. This means that whether a `Future` spawns or not is "leaked" all throughout its call stack, which makes it hard to change down the line.

Reply:
Ok, so now I'm confused about what default means. What you're suggesting is that this function spawns with the 'current' executor (which I'd be fine with), but not with the 'default' executor.

'current' meaning, the executor which is currently associated with context.
'default' meaning, the executor which is some form of global default.

I'm against a function using a global default. I'm fine with a function using whatever executor is currently stored inside context, which by definition, can't be forced into a single global default.

Reply:
@ahmedcharles `ctx.spawn(fut)` will spawn on the `Spawn` currently associated with the context. `my_local_pool.spawn_local(fut)` will spawn onto the local pool. There is no global executor in this proposal.